rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -73,28 +74,32 @@ class AsyncParametersClient
     std::function<void(
       std::shared_future<std::vector<rclcpp::parameter::ParameterVariant>>)> callback = nullptr)
   {
-    std::promise<std::vector<rclcpp::parameter::ParameterVariant>> promise_result;
-    auto future_result = promise_result.get_future().share();
+    auto promise_result =
+      std::make_shared<std::promise<std::vector<rclcpp::parameter::ParameterVariant>>>();
+    get_parameters_promises_.insert(promise_result);
+    auto future_result = promise_result->get_future().share();
 
     auto request = std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
     request->names = names;
 
     get_parameters_client_->async_send_request(
       request,
-      [&names, &promise_result, &future_result, &callback](
+      [names, promise_result, &future_result, &callback, this](
```

Comment:
I made a note about the same thing in my branch that I pushed (which was when me and @esteve were hacking on it together):

https://github.com/ros2/rclcpp/commit/c6afe5c4d858a12f188c4c5334b9cef5f52977f6#diff-3d5a5b02e888be31f79e8c934aab2847R163

I'm not sure what to do about it, but it at least deserves a TODO so we don't forget about it. Maybe it should be handled by a weak pointer and `std::enable_shared_from_this`?